[#8] 회원가입 인증 구현

hoa0217 opened the pull request:
## 개요
회원가입 후 이메일 인증 기능을 구현하였으며, 간단한 타임리프 회원가입화면을 만들었습니다.

## 작업사항
- UserController : 인증api 추가
- UserService : 회원가입메서드 수정, 인증코드 생성 및 저장메서드 추가
- UserDao : 회원상태업데이트 메서드 추가
- UserDao, UserService 테스트케이스 추가
- 이메일 Event 및 EventHandler 생성
- Join화면 생성

## 변경로직
- 기존엔 회원가입 후 바로 DB에 저장하였지만 UNAUTHORIZED라는 상태를 추가하여 인증단계를 거치도록 로직을 변경하였습니다.
#### 회원가입 요청 순서
1. 폼에 정보 입력 후 회원가입요청
2. 인증코드생성 및 session에 저장
3. 인증링크담긴 메일 사용자에게 발송
4. UNAUTHORIZED상태로 사용자정보 DB에 저장
#### 회원가입 인증 순서
1. 사용자가 이메일의 인증링크 클릭 (인증 API 호출)
2. 넘어온 인증코드와 seesion에 담긴 인증코드 비교
3. 동일할 시 NORMAL상태로 사용자정보 업데이트

hoa0217 committed:
feat: "메일 전송 환경 구성"

- gradle 의존성 추가
- 메일관련 application.properties 추가
- MailService, MailServiceTest 추가
- 실제로 메일이 발송되는지를 테스트하기위해 SpringBootTest를 진행함.

Related: #8

hoa0217 committed:
feat: "메일 전송 이벤트 생성"

- mail 패키지 생성
- 메일 이벤트 및 이벤트 핸들러 생성
- 실제 전송할 메일 내용 작성

Related: #8

hoa0217 committed:
refactor: "UNAUTHORIZED Enum 추가"

- 아직 메일 인증을 하지 않았을 때 상태값을 추가

Related: #8

hoa0217 committed:
refactor: "JoinRequest Status 변경"

- 아직 메일 인증을 하지 않았을 때 상태값으로 변경

Related: #8

hoa0217 committed:
feat: "회원가입 시 필요한 로직 추가"

- 회원가입로직 추가
  1. 인증코드 생성
  2. 인증메일이벤트 호출
  3. 인증코드 세션저장
- 회원가입테스트케이스 추가작성
- 임시 인증 api 작성

Related: #8

hoa0217 committed:
feat: "회원가입 인증 로직 추가"

- 세션에 담아뒀던 인증코드 검증 로직 추가
- 회원 상태 업데이트 로직 추가
- 인증요청 테스트케이스 작성

Related: #8

hoa0217 committed:
feat: "회원가입 화면 생성"

- gradle 타임리프 의존성 추가
- UserViewController 추가 (단순 뷰 리턴용)
- 간단한 회원가입 타임리프 화면 생성

Related: #8

hoa0217 committed:
refactor: "UserService테스트 필요없는 given 제거"

- userDao는 mock객체이므로 굳이 doNothing을 지정해주지 않아도 아무것도 안함.

Related: #8

hoa0217 committed:
fix: "dev merge"

- UserServiceImplTest conflict 해결

Related: #8

f-lab-fi commented on src/main/java/com/smart/mail/event/MailEvent.java:
```diff
@@ -0,0 +1,16 @@
+    this.email = email;
+    this.authCode = authCode;
+  }
+}
```
MailEvent라는 이름이 제가 느끼기에는 범용적이면서 직관적이지 않은 이름이라고 생각됩니다.
현재 이 객체는 본인 이메일 확인을 위한 인증 코드를 발송하는 이벤트에 활용되고 있는걸로 보이는데요,
만약 나중에 '메일을 사용하지만, 인증 메일이 아니라 단순히 애플리케이션에서 발생하는 알림 메일이 오는 경우'에 해당하는 메일 이벤트 객체를 정의해야한다면 어떨까요?
그럼 그 이벤트 객체의 이름을 뭘로 해야할지 고민할거고, 결국 이 객체의 이름까지 함께 바꿔야할 가능성이 높습니다.
차라리 이 객체의 이름을 Mail**Auth**Event 같이 조금 더 명확한 이름으로 바꾸는게 어떨까 싶습니다. :)

hoa0217 replied:
넵 맞는 말씀이십니다. 수정하겠습니다!

f-lab-fi commented on src/main/java/com/smart/mail/service/Implement/MailServiceImpl.java:
```diff
@@ -0,0 +1,42 @@
+package com.smart.mail.service.Implement;
```
패키지 네이밍 컨벤션이 이상한 것 같습니다.

**I**mplement 로 i가 대문자인데, 소문자가 되는게 적절한거 아닌지 확인 부탁드립니다.

hoa0217 replied:
넵 맞는 말씀이십니다. 수정하겠습니다!

f-lab-fi commented on src/main/java/com/smart/mail/event/MailEventHandler.java:
```diff
@@ -0,0 +1,20 @@
+import org.springframework.stereotype.Component;
+
+@Component
+public class MailEventHandler{
+
+  private final MailService mailService;
+
+  public MailEventHandler(MailService mailService) {
+    this.mailService = mailService;
+  }
+
+  @EventListener
+  public void eventHandle(MailEvent event) {
+    mailService.sendAuthMail(event.getEmail(), event.getAuthCode());
+  }
+}
```
인증 메일 발송에 대해 별도로 이벤트로 처리를 하셨는데요, 이런 부분을 왜 직접 메서드를 호출하지 않고 이벤트를 발행하여 처리했는지 물어본다면 뭐라고 답변하시겠나요~?

hoa0217 replied:
이벤트 발행 목적은 회원가입 시 user를 테이블에 저장하는 로직과 메일 발송에 대한 관심사 및 트랜잭션을 분리하고싶었습니다.
- 그런데 현재 트랜잭션 분리를 해놓지는 않아서 추가해놓겠습니다!

그리고 MailService를 직접 의존하면 나중에 SMS발송확장기능까지 고려했을 때 의존하는 Service가 너무 많아질 수 도 있다는 생각이 들었습니다.

f-lab-fi replied:
@hoa0217 

넵 좋습니다. 지금 말씀하신 방향성 대로라면 적절해보입니다.
종종 이벤트를 활용할 때 이벤트를 잘못 사용하는 경우도 있는데, 논리적으로나 트랜잭션으로나 분리되면 안되는 일련의 로직을 이벤트를 활용하는 경우가 있습니다. 그냥 생각나는 예시를 하나 들어보면

- 어떤 사용자가 모임을 생성한다.
- 그 모임의 주인을 그 사용자로 설정한다

라는 일련의 동작이 있을 때 이 두 동작은 논리적으로 하나이면서, 트랜잭션 상에서도 하나로 묶여야하는 동작입니다. 그런데 이런 동작을 이벤트를 발행하는 것으로 처리하여 논리적인 연관 관계를 희미하게 만들고 트랜잭션 상으로도 분리시켜 버리는 경우 이벤트를 잘못 활용하는 것일 수도 있습니다.

hoa0217 replied:
자세한 예시 감사합니다!! 일련의 동작에 대한 논리적 관계를 잘 생각하고 써야겠네요. 이해가 한번에 됐습니다. 🙂

f-lab-fi commented on src/main/java/com/smart/mail/service/Implement/MailServiceImpl.java:
```diff
@@ -0,0 +1,42 @@
+/**
+ * JavaMailSender : 스프링의 MailSender를 확장한 인터페이스이며 구현체는 JavaMailSenderImpl이다.
+ */
+@Service
+public class MailServiceImpl implements MailService {
+
```
MailService를 구현체를 바로 사용하지 않고, MailService라는 인터페이스를 둠으로써 어떤 이점이 생기는건가요?

hoa0217 replied:
service layer에서 interface와 구현체를 분리하는 습관이 있어 습관적으로 코딩한거같습니다.

이점은 메일서비스구현체를 변경하거나 확장하더라도 이벤트핸들러 코드에는 영향이 없다는 점입니다. 

그런데 사실 스스로 코딩하면서 정말 이러한 이점이 있을까? 싶은 의문이 들기는 합니다. 실제로 구현체를 확장해본 경험은 없어서.. 😅

f-lab-fi replied:
@hoa0217 

네 ㅎㅎ 일단 이점은 잘 이야기 해주셨는데, 이야기하신 것처럼 실제 이러한 이점이 있는지 고민해보셔야합니다.
이와 관련해서는 OCP(개방-폐쇄 원칙)나 DIP(의존 역전 원칙)를 생각해봐야하는데, 이와 관련해서는 패키지를 나누는 방법에 대해서 다다음 멘토링 시간 정도에 이야기를 한번 하려고 합니다.
지금은 너무 신경쓰지 마시고, 우선 그대로 두셔도 될 것 같습니다.

f-lab-fi commented on src/main/java/com/smart/mail/service/Implement/MailServiceImpl.java:
```diff
@@ -0,0 +1,42 @@
+
+    mailMessage.setTo(email);
+    mailMessage.setSubject("회원가입 인증 메일입니다.");
+    StringBuilder sb = new StringBuilder();
+    sb.append("[회원가입 인증]\n");
+    sb.append("아래 링크를 클릭하시면 회원가입 인증이 완료됩니다\n");
+    sb.append("http://localhost:8080/api/v1/user/join-auth?");
+    sb.append("email=" + email);
+    sb.append("&authCode=" + authCode);
```
StringBuilder로 문자열을 조합하면 가독성이 떨어지기 마련인데요, 혹시 다른 방법은 없을까요?

hoa0217 replied:
String에 +로 조합해주는것 보단 가독성있다고 생각하여 StringBuilder를 사용했는데 더 좋은 방법이 있는지 찾아보겠습니다.

f-lab-fi commented on src/main/java/com/smart/mail/service/Implement/MailServiceImpl.java:
```diff
@@ -0,0 +1,42 @@
+    StringBuilder sb = new StringBuilder();
+    sb.append("[회원가입 인증]\n");
+    sb.append("아래 링크를 클릭하시면 회원가입 인증이 완료됩니다\n");
+    sb.append("http://localhost:8080/api/v1/user/join-auth?");
```
이 링크는 개발자가 직접 로컬에서 테스트 할 때는 괜찮겠지만, 진짜 사용자에게 발송된다면 정상동작하지 않을겁니다.
현재에는 당연히 배포된 서버가 없기 때문에 정상적인 링크를 넣을 수 없겠지만, 실제 서비스가 배포된 상황을 가정하면 이 링크 주소를 어떻게 관리해볼 수 있을까요?

+) 이렇게 코드 내에 들어가 있어도 될지?

hoa0217 replied:
실제 서비스가 배포된다면 개발과 운영이 분리된 spring설정파일에 넣어 그 값을 가져오는게 좋을거라는 생각이 듭니다..

혹시 현업에서는 어떤 방식을 사용하는지 여쭈어봐도 괜찮을까요?

f-lab-fi replied:
넵 당연히 설정 파일에 넣어서 배포되어야합니다.
말씀해주신 것처럼 개발(dev), 운영(prod) 환경이 별도로 운영되는 경우 서로 다른 도메인 주소를 가지게 될텐데, 그런 상황을 가정해서라도 별도의 profile에 따른 설정 파일이 사용되는 방식을 사용하는게 좋습니다.

> 혹시 현업에서는 어떤 방식을 사용하는지 여쭈어봐도 괜찮을까요?

-> Spring Profile을 활용해서 나누는게 일반적입니다~

f-lab-fi commented on src/main/java/com/smart/mail/service/Implement/MailServiceImpl.java:
```diff
@@ -0,0 +1,42 @@
+    javaMailSender.send(makeAuthMail(email, authCode));
+  }
+
+  @Override
+  public SimpleMailMessage makeAuthMail(String email, String authCode) {
```
이 메서드가 public 일 이유가 있을까요?

hoa0217 replied:
`makeAuthMail` 메서드 말씀하시는걸까요? 이 부분은 굳이 public일 필요가 없는거같습니다. 수정하겠습니다!

f-lab-fi commented on src/main/java/com/smart/user/service/Implement/UserServiceImpl.java:
```diff
@@ -26,9 +39,33 @@ public void join(UserDto.JoinRequest request) {
+  public void saveAuthCode(String email, String authCode) {
+    session.setAttribute(email, authCode);
+    session.setMaxInactiveInterval(60 * 60 * 24); // TTL 24시간
+  }
```
이 두 메서드 역시 public일 이유가 없어보이는데 public인 이유가 있을까요!?

hoa0217 replied:
아니요 수정하겠습니다!

f-lab-fi commented on src/main/java/com/smart/user/service/Implement/UserServiceImpl.java:
```diff
@@ -26,9 +39,33 @@ public void join(UserDto.JoinRequest request) {
+    return UUID.randomUUID().toString();
+  }
+
+  @Override
+  public void saveAuthCode(String email, String authCode) {
+    session.setAttribute(email, authCode);
+    session.setMaxInactiveInterval(60 * 60 * 24); // TTL 24시간
+  }
+
+  @Override
+  public void verifyAuthCode(String email, String authCode) {
+    if(authCode.equals((String) session.getAttribute(email))){
+      session.removeAttribute(email);
+      userDao.updateUserStatus(email, Status.NORMAL);
+    }
+    else throw new IllegalAuthCodeException();
+  }
```
혹시 이 authCode를 세션에 저장하고 조회해오는 이유가 있을까요?
다른 방법을 생각해보면 User 테이블에 저장하는 방법도 있었을 것 같은데, 이렇게 구현하신 이유가 궁금합니다. :)

hoa0217 replied:
authCode는 인증할 때 한번만 사용되고 그 뒤로는 저장하고 있을 필요가 없는 데이터이기 때문에 테이블에 저장하는 것은 비효율적이라고 생각했습니다.

고려했던 방법으론 session과 redis였는데, session이 일단 빠르게 구현 가능하므로 채택하였습니다.

하지만 session을 사용하는 로직이 추가된다면 서로 영향을 받을 수 있기 때문에 추후 redis로 변경하는 것이 맞다고 생각합니다.

f-lab-fi replied:
@hoa0217 

넵 좋습니다~
근데 redis를 사용하는 쪽으로 바꾸는 것 관련해서는 'session을 사용하는 로직이 추가된다면 서로 영향을 받을 수 있기 때문에' 라기보다는 고려해볼만한 포인트가 2가지 있습니다.

- 인증 요청이 계속 쌓이면 session이 계속 쌓이는 것은 아닌지? 결국 이게 메모리를 차지하는데, 24시간 동안이나 기록된다는 것
- 웹 애플리케이션 서버가 여러개 떠서 로드밸런싱 되는 환경에서는 각 서버간 session을 공유해줘야하는데, 이걸 어떻게 해결할 수 있을지?

위 2가지 항목이 redis로 변경해보는 이유가 될 수 있을 것 같습니다.

hoa0217 replied:
헉....제가 2번째 포인트는 전혀 생각하지 못하고 있었습니다. 
말씀해주신 부분 고려하면 서비스 배포전엔 redis로 무조건 변경해야겠네요!! 감사합니다!!

f-lab-fi commented on src/main/resources/templates/user/join.html:
```diff
@@ -0,0 +1,55 @@
+    <script src="https://code.jquery.com/jquery-3.6.0.min.js"></script>
+</head>
+<script th:inline="javascript">
+    function send(){
+        const email =$("#email").val();
+        const password =$("#password").val();
+        const name =$("#name").val();
+        const nickname =$("#nickname").val();
+
+        const sendDTO = {
+            email:email,
+            password:password,
+            name:name,
+            nickname:nickname
+        };
+        $.ajax({
+            url: "/api/v1/user/join",
+            data: JSON.stringify(sendDTO),
+            contentType : 'application/json',
+            type: 'POST',
+        }).done(function (data){
+            console.log(data);
+           alert('회원가입이 완료되었습니다. 메일함에 가서 인증링크를 클릭해주세요.');
+        });
+    }
```
이거 혹시 jQuery 없이 똑같은 로직 작성하실 수 있으신가요~?

hoa0217 replied:
음 json 데이터 형식으로 보내려고  jQuery를 사용해 변환했습니다. form데이터 형식을 보낸다면 JQuery없이 작성 가능합니다!

아니면 혹시 물어보신 의도가 JQuery없이 json으로 변환해서 보낼 수 있냐는 말씀일까요??

f-lab-fi replied:
@hoa0217 

넵 제가 이야기드린건 jQuery 없이 비동기적으로(AJAX) JSON이 담긴 Body를 요청으로 보낼 수 있는지 여쭤본 거였습니다~

hoa0217 replied:
jQuery없이 작성해본적은 없어서 방법을 찾아봤는데 `Fetch API`를 통해 비동기적으로 통신할 수 있네요 ㅎㅎ 수정해보겠습니다!

f-lab-fi commented on src/test/java/com/smart/mail/service/Implement/MailServiceImplTest.java:
```diff
@@ -0,0 +1,42 @@
+import org.springframework.mail.SimpleMailMessage;
+import org.springframework.mail.javamail.JavaMailSender;
+
+@ExtendWith(MockitoExtension.class)
+class MailServiceImplTest {
+
+  @Mock
+  JavaMailSender javaMailSender;
+
+  @InjectMocks
+  MailServiceImpl mailService;
+
+  final String MAIL = "test@test.com";
+  final String AUTH_CODE = "authCode";
+
+  @DisplayName("회원가입 인증 메일을 전송한다.")
+  @Test
+  public void 인증메일전송() {
+    ArgumentCaptor<SimpleMailMessage> valueCapture = ArgumentCaptor
+        .forClass(SimpleMailMessage.class);
+    doNothing().when(javaMailSender).send(valueCapture.capture());
+
+    mailService.sendAuthMail(MAIL, AUTH_CODE);
+
+    verify(javaMailSender).send(any(SimpleMailMessage.class));
+    Assertions.assertEquals(MAIL, valueCapture.getValue().getTo()[0]);
+  }
+}
```
저라면 사실 이런 테스트는 작성하지 않았을 것 같습니다.
이 테스트가 실제로 테스트하고 있는게 어떤건지 한번 생각해보시죠. :)

- 이 테스트가 무엇을 검증하고 있는가?
- 이 테스트가 깨지는(실패하는) 순간은 무엇을 의미하는가, 코드를 수정하는 과정에서 코드를 잘못 수정해서 이 테스트가 깨질 가능성이 높을지?

hoa0217 replied:
- 이 테스트는 실제 메일 전송 구현체인 `javaMailSender`의 메일전송(`send`) 메서드의 파라미터가 올바르게 들어갔는지 검증하는 테스트입니다.
- 깨지는 순간은.. SimpleMailMessage를 잘못 만들어졌을 때 깨질거같습니다. SimpleMailMessage를 만드는 `makeAuthMail` 메서드를 수정할 경우 깨질 가능성이 높아 보여요.

그렇다면 `makeAuthMail` 메서드의 테스트 코드를 작성하는게 적합할까요?

f-lab-fi replied:
앗 이건 제가 테스트 코드를 잘못 읽은 것 같습니다.
제일 마지막줄을 못봐서 그냥 javaMailSender.send가 아무 SimpleMailMessage를 받으면 테스트에 통과되는줄 알았네요.
valueCapture에 대한 assert 문을 하나 더 추가해서 AUTH_CODE를 본문에 포함하고 있는지도 검사하면 더 좋을 것 같습니다.

다만 한가지 이야기 드리고 싶은건, 정화님도 이야기하신 것처럼 실제로 이 테스트는 makeAuthMail에 대한 테스트에 가깝습니다. 테스트가 깨질 가능성도 말씀하신 것처럼 makeAuthMail가 SimpleMailMessage를 만드는 과정이 잘못 변경되었을 때 실패했을거구요.

그러나 해당 메서드는 제가 리뷰 드렸던 것처럼 public 보다는 private으로 바뀌는게 적절한 메서드입니다. public 메서드인 sendAuthMail를 통해 간접적으로 테스트 되는게 적절해보입니다. 문제는 이 테스트의 이름이 '회원가입 인증 메일을 전송한다.'인게 어색한겁니다. ㅎㅎ 테스트 이름이 실제 테스트 되는 것(makeAuthMail가 올바른 SimpleMailMessage를 만들어 내는 것)과 잘 매칭이 되지 않는게 문제인 것 같습니다.

hoa0217 replied:
맞네요 makeAuthMail은 private 메서드이기 때문에 직접적인 테스트 코드 작성이 불가능하네요..ㅋㅋㅋㅋ죄송합니다.

넵 authCode도 확인하는 assert문 추가 및 테스트 이름도 적절하게 변경하도록 하겠습니다.

자세한 피드백 감사합니다!! 😄

f-lab-fi commented on src/test/java/com/smart/mail/service/MailServiceTest.java:
```diff
@@ -0,0 +1,23 @@
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.context.SpringBootTest;
+
+@SpringBootTest
+class MailServiceTest {
+
+  @Autowired
+  private MailService mailService;
+
+  final String MAIL = "gjwjdghk123@naver.com";
+  final String AUTH_CODE = "authCode";
+
+  @DisplayName("실제 메일 발송 여부를 테스트한다.")
+  @Test
+  public void 메일전송() {
+    mailService.sendAuthMail(MAIL, AUTH_CODE);
+  }
+
+}
```
이 테스트는 https://github.com/f-lab-edu/s-mart/pull/9/files#r1146561197 와는 반대로 의미가 있다고 생각할만한 테스트 입니다.
메일 발송에 실패하는 경우에 대한 테스트도 추가해보시기 바랍니다.
**메일 발송 실패 -> 테스트 성공** 하는 테스트를 이야기 드리는겁니다. :)

hoa0217 replied:
넵 알겠습니다!

hoa0217 reviewed (commented):
리뷰에 대한 comment 작성

hoa0217 committed:
refactor: "리뷰를 반영한 리팩토링"

- public 메서드 ➡️ private 메서드
- 패키지명 Implement ➡️ implement
- MailEvent ➡️ MailAuthEvent
  - 직관적인 이름으로 변경

Related: #8

hoa0217 committed:
refactor: "메일이벤트 비동기처리 추가"

- 비동기 처리를 추가하여 회원가입 저장 로직과 메일 발송의 트랜잭션을 분리하였다.

Related: #8

hoa0217 committed:
refactor: "메일이벤트 트랜잭션 분리"

- `Propagation.REQUIRES_NEW` 옵션을 통해 회원가입 저장 로직과 메일 발송의 트랜잭션을 분리하였다.

- 코드의 가독성과 유지보수성을 위해 Service layer에 @Transactional 어노테이션 추가

> 이전에 바꾼건 thread만 분리한것..

Related: #8

hoa0217 committed:
refactor: "가독성을 위한 String Format 사용"

- 가독성을 위해 StringBuilder에서 Strig Format으로 변경하였습니다.

Related: #8

hoa0217 committed:
refactor: "mail url 스프링설정파일에 저장"

- 개발환경, 운영환경이 분리되었을 시 url이 변경되므로 스프링설정파일에 넣어 관리합니다.

Related: #8

hoa0217 committed:
refactor: "회원가입화면 jQuery 제거"

- jQuery 대신 javascript에 내장된 fetch api를 사용합니다.

Related: #8

hoa0217 committed:
refactor: "sendAuthMail 테스트 수정"

- 메일 전송보다는 올바른 메일을 생성하는지 여부가 테스트의 중점이기 때문에 테스트 이름을 변경했습니다.
- AUTH_CODE확인 Assert문을 추가하였습니다.

Related: #8

hoa0217 committed:
refactor: "메일전송실패 테스트 추가"

- mail 주소가 null일 경우 메일전송에 실패합니다.
- 올바른 mail주소가 아닐 경우 메일전송에 실패합니다.

Related: #8

f-lab-fi commented on src/main/java/com/smart/mail/event/MailEventHandler.java:
```diff
@@ -13,8 +16,12 @@ public MailEventHandler(MailService mailService) {
+  @TransactionalEventListener
+  // 성능을 위해 비동기로 처리한다(회원가입 스레드와 메일전송 스레드 분리)
+  @Async
+  public void eventHandle(MailAuthEvent event) {
```
이벤트 리스너 대신 Async로 바꾸셨군요~
혹시 이렇게 바꾸신 이유가 어떤건지 알 수 있을까요?!

hoa0217 replied:
회원정보 DB저장 Transaction이 commit된 후 Event를 Listen받기 위해 `@TransactionalEventListener`으로 바꿨습니다! 
> `TransactionPhase phase() default TransactionPhase.AFTER_COMMIT;`

이를 통해 트랜잭션은 분리하였으나, 회원가입transaction, 이메일전송 transaction이 하나의 스레드로 동작하여.. 메일 전송이 끝날 때까지 사용자에게 응답을 주지 못하더라구요.

실제로 운영되고 있는 서비스들을 보면 인증 메일이나 인증 문자 발송이 완료 될때까지 기다렸다 사용자에게 응답을 주기보단 비동기로 통신하는 경우를 많이 보았고 Async로 변경하였습니다!
